src/backend/distributed/commands/local_multi_copy.c

```diff
@@ -0,0 +1,238 @@
+ * we are going to be using binary format internally.
+ *
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "commands/copy.h"
+#include "catalog/namespace.h"
+#include "parser/parse_relation.h"
+#include "utils/lsyscache.h"
+#include "nodes/makefuncs.h"
+#include "safe_lib.h"
+#include <netinet/in.h> /* for htons */
+
+#include "distributed/transmit.h"
+#include "distributed/commands/multi_copy.h"
+#include "distributed/multi_partitioning_utils.h"
+#include "distributed/local_executor.h"
+#include "distributed/local_multi_copy.h"
+#include "distributed/shard_utils.h"
+
+/*
+ * Data size threshold to switch over the active placement for a connection.
+ * If this is too low, overhead of starting COPY commands will hurt the
+ * performance. If this is too high, buffered data will use lots of memory.
+ * 512KB is a good balance between memory usage and performance. Note that this
+ * is irrelevant in the common case where we open one connection per placement.
+ */
+#define LOCAL_COPY_SWITCH_OVER_THRESHOLD (1 * 512 * 1024)
+
+
+static int ReadFromLocalBufferCallback(void *outbuf, int minread, int maxread);
+static Relation CreateCopiedShard(RangeVar *distributedRel, Relation shard);
+static void AddSlotToBuffer(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest,
+							bool isBinary);
+
+static bool ShouldSendCopyNow(StringInfo buffer);
+static void DoLocalCopy(StringInfo buffer, Oid relationId, int64 shardId,
+						CopyStmt *copyStatement);
+static bool ShouldAddBinaryHeaders(StringInfo buffer, bool isBinary);
+
+/*
+ * localCopyBuffer is used in copy callback to return the copied rows.
+ * The reason this is a global variable is that we cannot pass an additional
+ * argument to the copy callback.
+ */
+StringInfo localCopyBuffer;
+
+/*
+ * ProcessLocalCopy adds the given slot and does a local copy if
+ * this is the end of copy, or the buffer size exceeds the threshold.
+ */
+void
+ProcessLocalCopy(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest, int64 shardId,
+				 StringInfo buffer, bool shouldSendNow)
+{
+	/*
+	 * Here we save the previous buffer, and put the local shard's buffer
+	 * into copyOutState. The motivation is to use the existing logic to
+	 * serialize a row slot into buffer.
+	 */
+	StringInfo previousBuffer = copyDest->copyOutState->fe_msgbuf;
+	copyDest->copyOutState->fe_msgbuf = buffer;
+
+	/* since we are doing a local copy, the following statements should use local execution to see the changes */
+	TransactionAccessedLocalPlacement = true;
+
+	bool isBinaryCopy = copyDest->copyOutState->binary;
+	AddSlotToBuffer(slot, copyDest, isBinaryCopy);
+
+	if (shouldSendNow || ShouldSendCopyNow(buffer))
+	{
+		if (isBinaryCopy)
+		{
+			AppendCopyBinaryFooters(copyDest->copyOutState);
+		}
+
+		DoLocalCopy(buffer, copyDest->distributedRelationId, shardId,
```

Comment:
Have you considered using `ExecSimpleRelationInsert()`? It probably has some performance disadvantages since it won't batch rows, but on the other hand with using COPY we are adding a deserialization overhead (we do tuple -> copy format -> tuple -> insert) which won't exist if we use `ExecSimpleRelationInsert()` (we do tuple -> insert. Although we need to do tuple -> copy format to ship row over connections, but the "copy format -> tuple" part is eliminated).

Maybe a quick a performance comparison can help us here.

Reply:
I havent because I didnt know there was such a method. We can probably do a performance comparison yeah

Reply:
It seems that using ExecSimpleRelationInsert is slower, I have modified the existing branch to use that and push it here: https://github.com/citusdata/citus/tree/localCopy_ExecSimpleRelationInsert

The relative part is here: https://github.com/citusdata/citus/blob/localCopy_ExecSimpleRelationInsert/src/backend/distributed/commands/local_multi_copy.c#L65-L88, whjich is called every time we receive a new slot from the copy.

I have done some runs:

with a reference table:
Copying 800K rows:
ExecSimpleRelationInsert: Time: 3856.462 ms (00:03.856)
This Branch: Time: 921.526 ms

Copying 15K rows:
ExecSimpleRelationInsert:  Time: 48.265 ms
This Branch: Time: 14.010 ms

With a distributed table:
Copying 70K rows: 
ExecSimpleRelationInsert:   Time: 188.988 ms
This branch: Time: 78.212 ms

The table definitions:
```
CREATE TABLE ref_table(a int);
CREATE TABLE dist_table(a int, b int);
```

I can do more benchmarks if there is a type of query that you guys think could be faster with ExecSimpleRelationInsert.

From the results I would say that this branch is significantly faster than ExecSimpleRelationInsert. So I would suggest to keep the implementation the same as this branch.

What do you think @marcocitus, @pykello?

Reply:
Thanks, makes a pretty clear case for the current approach.

Reply:
Can you move the operations that need to be done once in https://github.com/citusdata/citus/commit/e7089d7d7f0d33d8ef70d45b8bf3999ed165730a to `InitializeCopyShardState` and `CitusCopyDestReceiverShutdown` to make a more proper comparison?

For example, we don't need to do `heap_open()`, `heap_close()`, `GetShardLocalTableOid()`, `ExecOpenIndices()`, ... for each row. They only need to be done only once per shard.

Your approach might still be better, but the current `ExecSimpleRelationInsert` branch is not best way we could use it.

The only call we need to make per row seems to be `ExecSimpleRelationInsert()`, most other calls can be done once per shard.

Reply:
Yeah, you are right. I have added https://github.com/citusdata/citus/commit/c462ecbfd65154fd54fe139b37b0edeb1dbb70ee on top of that branch and the query times got better, but it still seems significantly slower compared to this branch.

So for example with 800K row insert we it went down to 2 seconds from 3.8 seconds but this branch is 0.9 seconds.

Or with a distributed table for 70K rows it went down to 150ms from 188ms but this branch is 78ms. I might do some more benchmarks. 

(Also note that as that branch is experimental I am only getting the minimal functionality there)

Reply:
Thanks for the benchmark, I think the current approach is preferable then.